spring-boot/autoconfigure/src/main/java/org/togglz/spring/boot/autoconfigure/TogglzAutoConfiguration.java

```diff
@@ -0,0 +1,251 @@
+import org.togglz.core.activation.DefaultActivationStrategyProvider;
+import org.togglz.core.manager.CompositeFeatureProvider;
+import org.togglz.core.manager.EnumBasedFeatureProvider;
+import org.togglz.core.manager.FeatureManager;
+import org.togglz.core.manager.FeatureManagerBuilder;
+import org.togglz.core.manager.PropertyFeatureProvider;
+import org.togglz.core.repository.StateRepository;
+import org.togglz.core.repository.cache.CachingStateRepository;
+import org.togglz.core.repository.composite.CompositeStateRepository;
+import org.togglz.core.repository.file.FileBasedStateRepository;
+import org.togglz.core.repository.mem.InMemoryStateRepository;
+import org.togglz.core.spi.ActivationStrategy;
+import org.togglz.core.spi.FeatureProvider;
+import org.togglz.core.user.NoOpUserProvider;
+import org.togglz.core.user.UserProvider;
+import org.togglz.spring.boot.autoconfigure.TogglzProperties.FeatureSpec;
+import org.togglz.spring.listener.TogglzApplicationContextBinderApplicationListener;
+import org.togglz.spring.security.SpringSecurityUserProvider;
+import org.togglz.spring.web.FeatureInterceptor;
+
+import com.github.heneke.thymeleaf.togglz.TogglzDialect;
+
+/**
+ * {@link EnableAutoConfiguration Auto-configuration} for Togglz.
+ *
+ * @author Marcel Overdijk
+ */
+@Configuration
+@ConditionalOnProperty(prefix = "togglz", name = "enabled", matchIfMissing = true)
+@EnableConfigurationProperties(TogglzProperties.class)
+public class TogglzAutoConfiguration {
+
+    @Bean
+    public TogglzApplicationContextBinderApplicationListener togglzApplicationContextBinderApplicationListener() {
+        return new TogglzApplicationContextBinderApplicationListener();
+    }
+
+    @Configuration
+    @ConditionalOnMissingBean(FeatureProvider.class)
+    protected static class FeatureProviderConfiguration {
+
+        @Autowired
+        private TogglzProperties properties;
+
+        @Bean
+        public FeatureProvider featureProvider() {
+            PropertyFeatureProvider provider = new PropertyFeatureProvider(properties.getFeatureProperties());
+            Class<? extends Feature>[] featureEnums = properties.getFeatureEnums();
+            if (featureEnums != null && featureEnums.length > 0) {
+                return new CompositeFeatureProvider(new EnumBasedFeatureProvider(featureEnums), provider);
+            } else {
+                return provider;
+            }
+        }
+    }
+
+    @Configuration
+    @ConditionalOnMissingBean(FeatureManager.class)
+    protected static class FeatureManagerConfiguration {
+
+        @Autowired
+        private TogglzProperties properties;
+
+        @Bean
+        public FeatureManager featureManager(FeatureProvider featureProvider, List<StateRepository> stateRepositories,
+                UserProvider userProvider, ActivationStrategyProvider activationStrategyProvider) {
+            StateRepository stateRepository = null;
+            if (stateRepositories.size() == 1) {
```

Comment:
Can the list be  `null`?

Reply:
No, that's a `@Bean`, so it's parameters are injected by Spring, using matching beans. In case of a collection:

> In case of a Collection or Map dependency type, the container autowires all beans matching the declared value type.

as stated in `@Autowired` javadocs. If no bean matches, that collection will be empty.